pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -2481,45 +2377,201 @@ func validateDisks(field *k8sfield.Path, disks []v1.Disk) []metav1.StatusCause {
-						} else if size&(size-1) != 0 {
-							return false, fmt.Sprintf("Provided size of %d is not a power of 2", size)
-						}
-						return true, ""
-					}
-					if sizeOk, reason := checkSize(customSize.Logical); !sizeOk {
-						causes = append(causes, metav1.StatusCause{
-							Type:    metav1.CauseTypeFieldValueInvalid,
-							Message: reason,
-							Field:   field.Index(idx).Child("blockSize").Child("custom").Child("logical").String(),
-						})
-					}
-					if sizeOk, reason := checkSize(customSize.Physical); !sizeOk {
-						causes = append(causes, metav1.StatusCause{
-							Type:    metav1.CauseTypeFieldValueInvalid,
-							Message: reason,
-							Field:   field.Index(idx).Child("blockSize").Child("custom").Child("physical").String(),
-						})
-					}
-				}
+				causes = append(causes, validateCustomBlockSize(customSize, field, idx)...)
 			}
 		}
 	}
 
 	return causes
 }
 
+func validatePciAddress(disk v1.Disk, field *k8sfield.Path, idx int) (causes []metav1.StatusCause) {
+	if disk.Disk == nil || disk.Disk.PciAddress == "" {
+		return
+	}
+
+	if disk.Disk.Bus != v1.DiskBusVirtio {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("disk %s - setting a PCI address is only possible with bus type virtio.", field.Child("domain", "devices", "disks", "disk").Index(idx).Child("name").String()),
+			Field:   field.Child("domain", "devices", "disks", "disk").Index(idx).Child("pciAddress").String(),
+		})
+	}
+
+	_, err := hwutil.ParsePciAddress(disk.Disk.PciAddress)
+	if err != nil {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("disk %s has malformed PCI address (%s).", field.Child("domain", "devices", "disks", "disk").Index(idx).Child("name").String(), disk.Disk.PciAddress),
+			Field:   field.Child("domain", "devices", "disks", "disk").Index(idx).Child("pciAddress").String(),
+		})
+	}
+
+	return
+}
+
+func validateBusSupported(disk v1.Disk, bus v1.DiskBus, diskType string, field *k8sfield.Path, idx int) (causes []metav1.StatusCause) {
+	if len(bus) == 0 {
+		return
+	}
+
+	if bus == "ide" {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: "IDE bus is not supported",
+			Field:   field.Index(idx).Child(diskType, "bus").String(),
+		})
+	} else {
+		buses := []v1.DiskBus{v1.DiskBusVirtio, v1.DiskBusSCSI, v1.DiskBusSATA, v1.DiskBusUSB}
+		validBus := false
+		for _, b := range buses {
+			if b == bus {
+				validBus = true
+			}
+		}
+		if !validBus {
+			causes = append(causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: fmt.Sprintf("%s is set with an unrecognized bus %s, must be one of: %v", field.Index(idx).String(), bus, buses),
+				Field:   field.Index(idx).Child(diskType, "bus").String(),
+			})
+		}
```

Comment:
I'm not a big fan of this loop and I think we can simply check the values explicitly like 
```suggestion
        if bus != v1.DiskBusVirtio || 
                bus != v1.DiskBusSCSI ||
                bus != v1.DiskBusSATA ||
                bus != v1.DiskBusUSB {
                causes = append(causes, metav1.StatusCause{
                        Type:    metav1.CauseTypeFieldValueInvalid,
                        Message: fmt.Sprintf("%s is set with an unrecognized bus %s, must be one of: %v", field.Index(idx).String(), bus, buses),
                        Field:   field.Index(idx).Child(diskType, "bus").String(),
                })
        }
```

Reply:
@alicefr I think the condition here should be `&&` if the bus is not equal to any specified buses.

Reply:
@zakisk you are perfectly right :)